tests/test_update_recover_script.py

```
@@ -0,0 +1,556 @@
+    return SCRIPT.read_text(encoding="utf-8")
+
+
+def _run_bash_harness(
+    script: str,
+    tmp_path: Path,
+    *,
+    check: bool = True,
+) -> subprocess.CompletedProcess[str]:
+    source = tmp_path / "update-recover-source.sh"
+    source.write_text(_script().replace('\nmain "$@"\n', "\n"), encoding="utf-8")
+    return subprocess.run(
```

Ruff (0.15.21)

[error] `subprocess` call: check for execution of untrusted input

(S603)